### ISSUE TITLE
URGENT: Unable to Upgrade Server Certificate, Expiration Date Closing In

### ISSUE THREAD

**Issue by JustinKaffenberger** [microsoft/service-fabric#935]
We have a very simple 5 node, x509 secured cluster that we are trying to upgrade the server certificate on. We have tried the following things:

#### Add `ThumbprintSecondary` to the configuration with the new certificate thumbprint.
The updated cluster configuration in this circumstance looks like the following:
```
{
  "Properties": {
    "GoalStateExpirationReminderInDays": 0,
    "Security": {
      "$id": "1",
      "CertificateInformation": {
        "$id": "2",
        "ClusterCertificate": {
          "Thumbprint": "69 4B 3E F7 9F 56 1C DD 0C 15 F9 63 57 4F 55 BC F0 C6 32 AC",
          "X509StoreName": "My"
        },
        "ServerCertificate": {
          "Thumbprint": "FB 11 E3 73 BF 4E 10 41 E6 9B 24 AB 12 1C 49 76 5C 40 5D 8A",
          "ThumbprintSecondary": "69 4B 3E F7 9F 56 1C DD 0C 15 F9 63 57 4F 55 BC F0 C6 32 AC",
          "X509StoreName": "My"
        },
        "ReverseProxyCertificate": {
          "Thumbprint": "FB 11 E3 73 BF 4E 10 41 E6 9B 24 AB 12 1C 49 76 5C 40 5D 8A",
          "X509StoreName": "My"
        },
        "ClientCertificateThumbprints": [
          {
            "IsAdmin": true,
            "CertificateThumbprint": "FB 11 E3 73 BF 4E 10 41 E6 9B 24 AB 12 1C 49 76 5C 40 5D 8A"
          }
        ]
      },
      "ClusterCredentialType": "X509",
      "ServerCredentialType": "X509"
    },
    "NodeTypes": [
      {
        "Name": "NodeType0",
        "ClientConnectionEndpointPort": 19000,
        "HttpGatewayEndpointPort": 19080,
        "ReverseProxyEndpointPort": 19081,
        "LeaseDriverEndpointPort": 19002,
        "ClusterConnectionEndpointPort": 19001,
        "ServiceConnectionEndpointPort": 19003,
        "ApplicationPorts": {
          "StartPort": 20001,
          "EndPort": 22024
        },
        "EphemeralPorts": {
          "StartPort": 49152,
          "EndPort": 65535
        },
        "IsPrimary": true
      }
    ],
    "AddonFeatures": [
      "BackupRestoreService"
    ],
    "FabricSettings": [
      {
        "Name": "Setup",
        "Parameters": [
          {
            "Name": "FabricDataRoot",
            "Value": "C:\\ProgramData\\SF"
          },
          {
            "Name": "FabricLogRoot",
            "Value": "C:\\ProgramData\\SF\\Log"
          }
        ]
      },
      {
        "Name": "BackupRestoreService",
        "Parameters": [
          {
            "Name": "SecretEncryptionCertThumbprint",
            "Value": "FB 11 E3 73 BF 4E 10 41 E6 9B 24 AB 12 1C 49 76 5C 40 5D 8A"
          }
        ]
      }
    ],
    "EnableTelemetry": true,
    "DiagnosticsStore": {
      "StoreType": "FileShare",
      "DataDeletionAgeInDays": 21,
      "Connectionstring": "C:\\ProgramData\\SF\\DiagnosticsStore",
      "IsEncrypted": false
    },
    "FabricClusterAutoupgradeEnabled": true
  },
  "Nodes": [
    {
      "NodeName": "Node_5",
      "NodeTypeRef": "NodeType0",
      "IPAddress": "SFNODE6",
      "FaultDomain": "fd:/dc1/r5",
      "UpgradeDomain": "UD5"
    },
    {
      "NodeName": "Node_3",
      "NodeTypeRef": "NodeType0",
      "IPAddress": "SFNODE4",
      "FaultDomain": "fd:/dc1/r3",
      "UpgradeDomain": "UD3"
    },
    {
      "NodeName": "Node_2",
      "NodeTypeRef": "NodeType0",
      "IPAddress": "SFNODE3",
      "FaultDomain": "fd:/dc1/r2",
      "UpgradeDomain": "UD2"
    },
    {
      "NodeName": "Node_1",
      "NodeTypeRef": "NodeType0",
      "IPAddress": "SFNODE2",
      "FaultDomain": "fd:/dc1/r1",
      "UpgradeDomain": "UD1"
    },
    {
      "NodeName": "Node_0",
      "NodeTypeRef": "NodeType0",
      "IPAddress": "SFNODE1",
      "FaultDomain": "fd:/dc1/r0",
      "UpgradeDomain": "UD0"
    }
  ],
  "ApiVersion": "06-2018",
  "Name": "dev",
  "ClusterConfigurationVersion": "7.0.0"
}
```

#### Replace `Thumbprint` outright with the new certificate thumbprint.

```
{
  "Properties": {
    "GoalStateExpirationReminderInDays": 0,
    "Security": {
      "$id": "1",
      "CertificateInformation": {
        "$id": "2",
        "ClusterCertificate": {
          "Thumbprint": "69 4B 3E F7 9F 56 1C DD 0C 15 F9 63 57 4F 55 BC F0 C6 32 AC",
          "X509StoreName": "My"
        },
        "ServerCertificate": {
          "Thumbprint": "FB 11 E3 73 BF 4E 10 41 E6 9B 24 AB 12 1C 49 76 5C 40 5D 8A",
          "ThumbprintSecondary": "69 4B 3E F7 9F 56 1C DD 0C 15 F9 63 57 4F 55 BC F0 C6 32 AC",
          "X509StoreName": "My"
        },
        "ReverseProxyCertificate": {
          "Thumbprint": "FB 11 E3 73 BF 4E 10 41 E6 9B 24 AB 12 1C 49 76 5C 40 5D 8A",
          "X509StoreName": "My"
        },
        "ClientCertificateThumbprints": [
          {
            "IsAdmin": true,
            "CertificateThumbprint": "FB 11 E3 73 BF 4E 10 41 E6 9B 24 AB 12 1C 49 76 5C 40 5D 8A"
          }
        ]
      },
      "ClusterCredentialType": "X509",
      "ServerCredentialType": "X509"
    },
    "NodeTypes": [
      {
        "Name": "NodeType0",
        "ClientConnectionEndpointPort": 19000,
        "HttpGatewayEndpointPort": 19080,
        "ReverseProxyEndpointPort": 19081,
        "LeaseDriverEndpointPort": 19002,
        "ClusterConnectionEndpointPort": 19001,
        "ServiceConnectionEndpointPort": 19003,
        "ApplicationPorts": {
          "StartPort": 20001,
          "EndPort": 22024
        },
        "EphemeralPorts": {
          "StartPort": 49152,
          "EndPort": 65535
        },
        "IsPrimary": true
      }
    ],
    "AddonFeatures": [
      "BackupRestoreService"
    ],
    "FabricSettings": [
      {
        "Name": "Setup",
        "Parameters": [
          {
            "Name": "FabricDataRoot",
            "Value": "C:\\ProgramData\\SF"
          },
          {
            "Name": "FabricLogRoot",
            "Value": "C:\\ProgramData\\SF\\Log"
          }
        ]
      },
      {
        "Name": "BackupRestoreService",
        "Parameters": [
          {
            "Name": "SecretEncryptionCertThumbprint",
            "Value": "FB 11 E3 73 BF 4E 10 41 E6 9B 24 AB 12 1C 49 76 5C 40 5D 8A"
          }
        ]
      }
    ],
    "EnableTelemetry": true,
    "DiagnosticsStore": {
      "StoreType": "FileShare",
      "DataDeletionAgeInDays": 21,
      "Connectionstring": "C:\\ProgramData\\SF\\DiagnosticsStore",
      "IsEncrypted": false
    },
    "FabricClusterAutoupgradeEnabled": true
  },
  "Nodes": [
    {
      "NodeName": "Node_5",
      "NodeTypeRef": "NodeType0",
      "IPAddress": "SFNODE6",
      "FaultDomain": "fd:/dc1/r5",
      "UpgradeDomain": "UD5"
    },
    {
      "NodeName": "Node_3",
      "NodeTypeRef": "NodeType0",
      "IPAddress": "SFNODE4",
      "FaultDomain": "fd:/dc1/r3",
      "UpgradeDomain": "UD3"
    },
    {
      "NodeName": "Node_2",
      "NodeTypeRef": "NodeType0",
      "IPAddress": "SFNODE3",
      "FaultDomain": "fd:/dc1/r2",
      "UpgradeDomain": "UD2"
    },
    {
      "NodeName": "Node_1",
      "NodeTypeRef": "NodeType0",
      "IPAddress": "SFNODE2",
      "FaultDomain": "fd:/dc1/r1",
      "UpgradeDomain": "UD1"
    },
    {
      "NodeName": "Node_0",
      "NodeTypeRef": "NodeType0",
      "IPAddress": "SFNODE1",
      "FaultDomain": "fd:/dc1/r0",
      "UpgradeDomain": "UD0"
    }
  ],
  "ApiVersion": "06-2018",
  "Name": "dev",
  "ClusterConfigurationVersion": "7.0.0"
}
```

#### Errors

The resulting errors during the upgrade are non-deterministic, but always revolve around an `FABRIC_E_SERVER_AUTHENTICATION_FAILED` error occurring on a random node during the upgrade process. For example:

```
'System.Hosting' reported Error for property 'FabricDownload:7.0.464.9590:22'.
There was an error during download. Failed to download 'WindowsFabricStore\ClusterManifest.22.xml' from ImageStore. Error 'FABRIC_E_SERVER_AUTHENTICATION_FAILED'. Check if the file or folder is present in ImageStore by using 'Get-ServiceFabricImageStoreContent -RemoteRelativePath WindowsFabricStore\ClusterManifest.22.xml'.
```

#### Answers to Obvious Questions
- Yes, the certificate in question is installed on all of the nodes, in all of the recommended stores, and NETWORK SERVICE has access to the certificate.
- Yes, I've tried adding fresh nodes and removing nodes exhibiting the error. The error still persists on new, cleanly created nodes.
- Yes, we've already performed the certificate rollover on the `ClusterCertificate` with no problems.
- Yes, if we run `Get-ServiceFabricImageStoreContent -RemoteRelativePath WindowsFabricStore\ClusterManifest.22.xml` on the node exhibiting the error, it runs successfully.

#### In Conclusion
We are effectively hosed unless a workaround or fix is provided this issue. The server certificate, at this moment in time, is impossible to upgrade. We would greatly appreciate any insight into this issue.

**Comment by Phineas019** [comment]
Hi,

Im facing the same issue, Can you please tell me how you fixed it.

**Comment by Phineas019** [comment]
No the issue is they are not able to download the image store file, with in the given timeout you have manually copy the files from one machine to other fixed the issue, Since I was doing it on production i didn't tried to explore much maybe I'll create dedicated R&D environment to know more about this issue ®

**Comment by JustinKaffenberger** [comment]
As it turns out there is a bug in Service Fabric related to rolling over certificates using thumbprints. While the MS support ticket is now closed (as they assisted us in finding a workaround for our specific situation), as far as I know the bug still exists and will hopefully be patched in the future. We are staying in contact with them around this issue since thumbprints are common in small, private deployment scenarios.

The workaround was quite obscure, and I'd recommend opening a ticket with MS if you need similar production support. I'll be sure to update this thread if they ever patch the issue.

**Comment by dragav** [comment]
@JustinKaffenberger which bug are you referring to, do you happen to have a GH issue number, or at an internal reference (bug description, contact info etc.)

I diffed your before/after configurations and they are identical. From your description, you attempted to replace _in place_ the thumbprint of a certificate, which is not supported/won't work. Deploying a configuration upgrade creates a partitioning of the cluster - nodes in upgraded domains will use the new settings, whereas un-upgraded nodes are on the initial/starting settings. When the changes in question are security settings, this reflects directly on to the mutual authentication of nodes and can cause loss of availability. For that reason, certificate changes are always orchestrated in 2 steps - in your case, to add a secondary thumbprint.

Certificate-based authentication in SF clusters is described [here](https://docs.microsoft.com/en-us/azure/service-fabric/cluster-security-certificates); there is a section dedicated to rotation of certs declared by thumbprint.

Note there may be, indeed, a bug - I'd still like to get more details, if you can share. If you'd rather not do so on a public forum, please email me directly - dragosav at microsoft etc.

**Comment by JustinKaffenberger** [comment]
@dragav yes we followed all of the docs at first, any other strange attempts were made by us out of desperation. We had a week long troubleshooting session with Microsoft Support in which we ended up at the conclusion that there was a bug. Would it be helpful if I gave you the ticket number or our support contact?

The other reason we landed on this 'bug' conclusion was that we were able to reproduce the issue on all of our clusters running on the latest SF runtime, however we had one cluster which was behind a few versions, and that cluster was able to upgrade via thumbprint rotation successfully.

**Comment by JustinKaffenberger** [comment]
@dragav This is the final mitigation comments provided by Microsoft Support:

> XXX cluster - 7.0.466 - Mitigated
> 1. generate the secondary cert using this powershell script - C:\Program Files\Microsoft SDKs\Service > Fabric\ClusterSetup\Secure\CertSetup.ps1
> 2. add the sec cert into the nodes using the below query(Install-ClusterCertificate).
> 3. add the secondary cert to all section like server, cluster and client using below commands.
>   But upgrade failed but this time they captured the error(refer below).
>   Get-ServiceFabricClusterConfiguration > NewClusterConfig.json
>   Start-ServiceFabricClusterConfigurationUpgrade -ClusterConfigPath NewClusterConfig.json
 
> 'System.Hosting' reported Error for property 'FabricDownload:7.0.466.9590:1'.
> There was an error during download. Failed to download 'WindowsFabricStore\ClusterManifest.1.xml' from ImageStore. Error > 'FABRIC_E_SERVER_AUTHENTICATION_FAILED'. Check if the file or folder is present in ImageStore by using 'Get-> ServiceFabricImageStoreContent -RemoteRelativePath WindowsFabricStore\ClusterManifest.1.xml'.


> Current Scenario: Cluster got stuck in upgrading status after adding new secondary cert over cluster config json file.  Once we > got into the situation. We should follow the below mitigation for now.
> Cause: This is a bug.
> Mitigation: Delete the new cert from all nodes. Once we deleted, upgrade will resume i believe and eventually it will be > succeeded and we can install new cert on all nodes. 
> Mitigation was successful.
> Resolution: It looks like an bug on cluster manifest XML generation code from standalone configuration file json.

**Comment by JustinKaffenberger** [comment]
In case the team isn't aware, this msftbot is marking things as 'status-olderthan1year' that are not older than one year.

**Comment by amenarde** [comment]
Hi @JustinKaffenberger, this matches a bug in certificate rotations in service fabric standalone clusters. This bug was identified and is fixed in 7.2. RTO. Unfortunately it doesn't look like the fix is captured in the Rel Notes. I will try to get this publicly noted. After upgrading to 7.2., cert rotations by thumbprint should smooth out. Please let us know if you are still having issues or need guidance for a cluster which cannot come onto 7.2. before rotation. Thanks

**Comment by JustinKaffenberger** [comment]
@amenarde thanks for the reply! We have since moved to CommonName based cert configuration but thumbprint-based is still really nice to have for edge (sometimes even offline altogether) deployments. I will be sure to try it out the next time I get a chance. Closing this for now.
